### ISSUE TITLE
Web UI becomes unusable under heavy load

### ISSUE THREAD

**Issue by ebr** [argoproj/argo-workflows#6223]
## Summary 

We often run many large workflows simultaneously: ~100 workflows with ~6000-8000 pods per workflow. Under these conditions, the Argo Server web interface fails with the following symptoms:
- rows fail to load - the "loading" animation (side scrolling) is displayed
- eventually the server shows a `Gateway timed out` error

This prevents us from being able to observe workflow status once the workload on the cluster reaches a certain size.

## Diagnostics

- K8s v1.20
- Argo Workflows v3.0.7
- Docker executor
- the web UI is served by the nginx ingress controller, fronted by AWS ELB
- actual workflow execution seems unaffected
- postgres DB isn't experiencing any unusual load
- Argo Server memory usage balloons to 45GB per replica (we tried with 1 to 3 replicas, attached screenshot is with 2 replicas; results are identical.
- Workflow controller does not exhibit any unusual behaviour
- Argo server log files are flooded with the following messages:

```
time="2021-06-25T13:23:23.326Z" level=warning msg="Non-transient error: upper: no more rows in this result set"
time="2021-06-25T13:23:23.326Z" level=error msg="finished streaming call with code Unknown" error="upper: no more rows in 
this result set" grpc.code=Unknown grpc.method=WatchWorkflows grpc.service=workflow.WorkflowService grpc.start_time="2021-06-25T13:15:59Z" grpc.time_ms=443903.97 span.kind=server system=grpc
time="2021-06-25T13:23:39.133Z" level=info msg="Alloc=572937 TotalAlloc=2151971522 Sys=910939 NumGC=29435 Goroutines=205"
time="2021-06-25T13:28:39.133Z" level=info msg="Alloc=294206 TotalAlloc=2157739650 Sys=910939 NumGC=29464 Goroutines=211"
time="2021-06-25T13:30:23.664Z" level=warning msg="Non-transient error: upper: no more rows in this result set"
time="2021-06-25T13:30:23.664Z" level=error msg="finished streaming call with code Unknown" error="upper: no more rows in this result set" grpc.code=Unknown grpc.method=WatchWorkflows grpc.service=workflow.WorkflowService grpc.start_time="2021-
06-25T13:23:40Z" grpc.time_ms=403471.44 span.kind=server system=grpc
time="2021-06-25T13:33:39.133Z" level=info msg="Alloc=290515 TotalAlloc=2169557577 Sys=910939 NumGC=29523 Goroutines=219"
time="2021-06-25T13:34:14.229Z" level=info msg="finished streaming call with code Canceled" error="rpc error: code = Canceled desc = context canceled" grpc.code=Canceled grpc.method=WatchWorkflows grpc.service=workflow.WorkflowService grpc.star
t_time="2021-06-25T13:31:47Z" grpc.time_ms=146437.19 span.kind=server system=grpc
```

![image](https://user-images.githubusercontent.com/2433541/123439770-0fb5fa00-d5a0-11eb-849c-2edd6381bbcf.png)

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by alexec** [comment]
The argo-server is a stateless application, so should happily run with just a few MB of memory.

Can you confirm which pages seems to have the problem? It's not clear to me.

Can I also ask you try v3.1.0? This includes #6004 which should reduce memory usage by reducing allocs when watching and listing workflows.

If that does not help, then I'll look to create an engineering build that enables a pprof endpoint. for capturing allocs/heap/profile etc.

**Comment by ebr** [comment]
The `/workflows/` pages are problematic - with or without either the namespace, or the `?limit=` query arg. I might be able to try 3.1.0 sometime next week, looking forward to it actually, but can't make any changes now.

I do have a bit more info though:
- restarting the server (scaling the deployment) did not help;
- RAM usage seems to grow only after we try to access one of the `/workflows/` pages
- I realized that the `argo_workflows` table in Postgres has grown to 7GB in size. After dropping the argo_* tables and restarting the workflow controller, the server UI loaded immediately
- we are able to consistently reproduce this; the server responsiveness and RAM usage seems to be directly correlated with the size of `argo_workflows` table;
- during my latest test I noticed something odd that I haven't reproduced yet. We deleted all workflows (server UI still unresponsive), checked `argo_workflows` table size, it was at 1.8GB. Scaled down both workflow controller and server, table size immediately dropped to 280MB without any other action from our end.

I'd be happy to make any additional observations/testing that don't require any significant changes for now, and try 3.1.0 or an engineering build at some point in the near future when we can afford it.

**Comment by ebr** [comment]
It seems that as the number of workflow nodes grows, the server's Postgres query `SELECT "uid", "version", "nodes" FROM "argo_workflows" WHERE ("clustername" = $1 AND "namespace" = $2)` blocks all further queries, including the ones from the workflow controller. This starts being noticeable (and eventually grinds to a halt) when the total number of workflow nodes (maybe only `Running`? - unconfirmed) in the namespace grows to tens of thousands. Also, this doesn't seem to be (as much of?) a problem if the UI is already open, but on the initial load (or refresh) it reliably explodes.

**Comment by alexec** [comment]
How did you testing with v3.1 go?

**Comment by ebr** [comment]
> How did your testing with v3.1 go?

Hoping to get to it next week - sorry, crazy time crunch right now (which is the reason for running these huge workflows in the first place :sweat_smile:

**Comment by alexec** [comment]
Should be fixed in `:latest` (v3.2)
